modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C

```diff
@@ -0,0 +1,126 @@
+    _coeff(declareModelData<std::vector<Real>>("_coeff")),
+    _power_matrix(declareModelData<std::vector<std::vector<unsigned int>>>("_power_matrix")),
+    _max_degree(declareModelData<unsigned int>("_max_degree", getParam<unsigned int>("max_degree")))
+{
+}
+
+void
+PolynomialRegressionTrainer::initialSetup()
+{
+  // Results VPP
+  _values_distributed = isVectorPostprocessorDistributed("results_vpp");
+  _values_ptr = &getVectorPostprocessorValue(
+      "results_vpp", getParam<std::string>("results_vector"), !_values_distributed);
+
+  // Sampler
+  _sampler = &getSamplerByName(getParam<SamplerName>("sampler"));
+
+  _n_dims = _sampler->getNumberOfCols();
+
+  // Initializing power matrix, using _max_degree+1 to tackle the indexing offset
+  // within generateTuple
+  _power_matrix =
+      StochasticTools::MultiDimPolynomialGenerator::generateTuple(_n_dims, _max_degree + 1);
+
+  _n_poly_terms = _power_matrix.size();
+
+  // Check if we have enough data points to solve the problem
+  if (_sampler->getNumberOfRows() <= _n_poly_terms)
+    mooseError("Number of data points must be greater than the number of terms in the polynomial.");
+
+  // Resize _coeff, _matrix, _rhs to number of sampler columns
+  _coeff.resize(_n_poly_terms);
+  _matrix.resize(_n_poly_terms, _n_poly_terms);
+  _rhs.resize(_n_poly_terms);
+}
+
+void
+PolynomialRegressionTrainer::initialize()
+{
+  // Check if results of samples matches number of samples
+  __attribute__((unused)) dof_id_type num_rows =
+      _values_distributed ? _sampler->getNumberOfLocalRows() : _sampler->getNumberOfRows();
+  mooseAssert(num_rows == _values_ptr->size(),
+              "Sampler number of rows does not match number of results from vector postprocessor.");
+}
+
+void
+PolynomialRegressionTrainer::execute()
+{
+  // Offset for replicated/distributed result data
+  dof_id_type offset = _values_distributed ? _sampler->getLocalRowBegin() : 0;
+
+  // Loop over samples
+  for (dof_id_type p = _sampler->getLocalRowBegin(); p < _sampler->getLocalRowEnd(); ++p)
+  {
+    std::vector<Real> data = _sampler->getNextLocalRow();
+
+    for (unsigned int i = 0; i < _n_poly_terms; ++i)
+    {
+      std::vector<unsigned int> i_powers(_power_matrix[i]);
+
+      Real i_value(1.0);
+      for (unsigned int ii = 0; ii < data.size(); ++ii)
+        i_value *= pow(data[ii], i_powers[ii]);
+
+      for (unsigned int j = 0; j < _n_poly_terms; ++j)
+      {
+        std::vector<unsigned int> j_powers(_power_matrix[j]);
+
+        Real j_value(1.0);
+        for (unsigned int jj = 0; jj < data.size(); ++jj)
+          j_value *= pow(data[jj], j_powers[jj]);
+
+        _matrix(i, j) += i_value * j_value;
+      }
+
+      _rhs(i) += i_value * (*_values_ptr)[p - offset];
+    }
+  }
+}
+
+void
+PolynomialRegressionTrainer::finalize()
+{
+  gatherSum(_matrix.get_values());
+  gatherSum(_rhs.get_values());
+
+  DenseVector<Real> sol;
+  _matrix.lu_solve(_rhs, sol);
+  _coeff = sol.get_values();
```

Comment:
I just wanted to leave note here for @aeslaughter and @cbolisetti in regards to what we are leaning in our MIT course. In this model we have a quadratic objective function:
```
f(a) = \sum_{i=1}^n (a_0 + a_1 x_{1,i} + a_2 x_{2,i} + ... - y_i)^2 = ||Xa - y||^2
```
So we are trying to do min_a [f(a)]. Taking the gradient, we get:
```
\nabla f(a) = X^T X a - X^T y = 0 ---> (X^T X) a = X^T y
```
Which we can solve as linear system of equations:
```
a = (X^T X)^-1X^T y
```
This is exactly what Peter has implemented here. However, we can actually solve this with gradient descent like:
```
a^(k+1) = a^(k) - eta \nabla f(a^(k))
```
With newton iteration, we actually get back the linear system:
```
a^(k+1) = a^(k) - [Hf(a^(k))]^-1 \nabla f(a^(k)) = a^(k) - (X^T X)^-1( X^T X a^(k) - X^T y) = (X^T X)^-1 X^T y
```
So my thinking is that we can create a generalized optimization class/utility that performs a version of gradient descent where the surrogate provides a function for `f(a)` and `\nabla f(a)`. `Hf(a)` could be defined explicitly by the surrogate, computed with AD, or simply left out.

Thoughts?

Reply:
We have an SGD implementation [here](https://github.com/idaholab/mastodon/blob/devel/src/utils/MastodonUtils.C) in `maximizeLogLikelihood` (but it needs to be its own util), that can be used as a starting point..